docs/development-guide/extending-gaffer.md

```diff
@@ -0,0 +1,217 @@
+
+### Writing an Operation
+
+In Gaffer we use the term Operation to describe some component of a
+manipulation/query affecting a graph, these are chained together in `OperationChains`
+and this is how we execute manipulations/queries to our graphs. Whilst
+there are a wealth of existing operations and their associated handlers that are
+provided out of the box with Gaffer; in this section we'll talk about how you
+might go about adding a new operation/handler with some simple examples.
+
+#### The Operation Interface
+
+All Operations in one form or another must adhere to and implement the [`Operation`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/operation/Operation.html)
+interface, this interface describes the building blocks an operation must be built from.
+
+One important distinction here is that an implementation of `Operation` should not
+contain the actual logic for an `Operation`, think of these as a task definition
+or a instruction manual for what an `Operation` should do. This then goes hand in
+hand with our `OperationHandler`(s) which define the actual functionality
+either generically or split into multiple handlers for different stores.
+
+#### Selecting an Interface
+
+In general most `Operation` implementations don't actually directly implement the
+`Operation` Interface, they actually implement one or more extended interfaces,
+for example the `Input` and `Output` interfaces extend on top of the core `Operation`
+interface and concordantly the `InputOutput` interface is a combination of these
+two to be used. These all live inside the Gaffer Core library under the
+`Operation` subdirectory (see the [component breakdown](./project-structure/components/operation.md)
+for more details).
+
+#### Implementing an Interface
+
+The example we'll use here is the
+[`ToSet`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/operation/impl/output/ToSet.html)
+Operation, this will takes an `Iterable` of `Elements` and converts it into a
+`Set`, a handy way to remove duplicate `Elements` if you're doing multiple hops
+and end up getting duplicates back.
+
+Taking a look at the [Javadoc](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/operation/impl/output/ToSet.html)
+for this class you should be able to see this class it implements both
+`InputOutput` (with an input type of `Iterable<? extends T>` and an output type
+of `Set<? extends T>` and `MultiInput<T>`). Within the class it also implements
+all the requisite methods eg, `shallowClone()`, `setInput()`, `getOptions()`
+etc.
+
+#### How to Write the Handler
+
+Using the same `ToSet` Operation from before, now that we have this definition
+of an operation, we need to actually tell Gaffer how to do it, we do this by
+defining either a single generic `Handler` and linking it to the Operation or in
+many cases we might want different implementations based on the store being used
+so we'd write a number of operation handlers and decide which one is needed at
+runtime.
```

Comment:
This is a really long sentence that isn't the clearest so could do with a re-visit.